File: PostProcessor.cpp
#include "PostProcessor.h"
#include <iostream>
#include <algorithm>

//класс ColorCorrectionEffect
//цветокоррекция мозаики на основе оригинального изображения
// (выравнивает средние значения цветовых каналов мозаики по оригиналу)
cv::Mat ColorCorrectionEffect::apply(const cv::Mat& mosaic, const cv::Mat& original) {
    cv::Mat corrected = mosaic.clone();
    cv::Mat originalResized;

    //приводим оригинальное изображение к размеру мозаики
    cv::resize(original, originalResized, mosaic.size());

    //вычисляем средние значения цветов для обоих изображений
    cv::Scalar mosaicMean = cv::mean(mosaic);
    cv::Scalar originalMean = cv::mean(originalResized);

    //разделяем изображение на цветовые каналы для независимой коррекции
    std::vector<cv::Mat> mosaicChannels, correctedChannels;
    cv::split(mosaic, mosaicChannels);

    //корректируем каждый цветовой канал отдельно
    for (int i = 0; i < mosaic.channels(); i++) {
        //коэффициент масштабирования для канала
        // +1e-5 для избежания деления на ноль
        double scale = originalMean[i] / (mosaicMean[i] + 1e-5);

        //применяем интенсивность эффекта к коэффициенту
        scale = 1.0 + (scale - 1.0) * intensity;

        //применяем коррекцию к каналу
        cv::Mat correctedChannel;
        mosaicChannels[i].convertTo(correctedChannel, CV_32F); //float для точных вычислений
        correctedChannel *= scale; //применяем масштабирование
        correctedChannel.convertTo(correctedChannel, mosaicChannels[i].type()); //возвращаем к исходному типу

        correctedChannels.push_back(correctedChannel);
    }

    //объединяем скорректированные каналы обратно в цветное изображение
    cv::merge(correctedChannels, corrected);
    return corrected;
}

//геттер эффекта цветокоррекции
std::string ColorCorrectionEffect::getName() const {
    return "color_correction";
}

//класс AlphaBlendEffectс
//смешивает мозаику с оригинальным изображением через альфа-канал
cv::Mat AlphaBlendEffect::apply(const cv::Mat& mosaic, const cv::Mat& original) {
    cv::Mat originalResized;
    //приводим оригинальное изображение к размеру мозаики
    cv::resize(original, originalResized, mosaic.size());

    //взвешенное сложение двух изображений
    cv::Mat blended;
    cv::addWeighted(mosaic, 1.0 - alpha, originalResized, alpha, 0, blended);

    return blended;
}

//геттер эффекта альфа-смешивания
std::string AlphaBlendEffect::getName() const {
    return "alpha_blend";
}

//класс SeamSmoothingEffect 
//сглаживает видимые швы между плитками мозаики(размытием)
cv::Mat SeamSmoothingEffect::apply(const cv::Mat& mosaic, const cv::Mat& original) {
    cv::Mat smoothed = mosaic.clone();

    //маска областей швов
    cv::Mat seamMask = createSeamMask(mosaic);

    //если нет швов для сглаживания, возвращаем исходное
    if (cv::countNonZero(seamMask) == 0) {
        return smoothed;
    }

    //параметры размытия
    int blurSize = 3 + static_cast<int>(intensity * 5);
    if (blurSize % 2 == 0) blurSize++;
    blurSize = std::max(3, std::min(blurSize, std::min(mosaic.rows, mosaic.cols)));

    if (blurSize < 3) return mosaic.clone();

    double sigma = intensity * 2.0;

    //размываем только швы
    cv::Mat blurred;
    cv::GaussianBlur(mosaic, blurred, cv::Size(blurSize, blurSize), sigma);

    //смешиваем размытую версию с оригиналом в областях швов
    double blendFactor = intensity * 0.7;

    //временное изображение для смешивания
    cv::Mat blended = mosaic.clone();
    cv::addWeighted(mosaic, 1.0 - blendFactor, blurred, blendFactor, 0, blended);

    //копируем сглаженные области швов из blended в результат
    blended.copyTo(smoothed, seamMask);

    return smoothed;
}

//геттер эффекта сглаживания швов
std::string SeamSmoothingEffect::getName() const {
    return "seam_smoothing";
}

//сеттер для размера сетки
void SeamSmoothingEffect::setGridSize(int gridSize) {
    this->gridSize = gridSize;
}

//бинарная маска областей швов между плитками мозаики
//определяет вертикальные и горизонтальные границы сетки мозаики
cv::Mat SeamSmoothingEffect::createSeamMask(const cv::Mat& mosaic) const {
    cv::Mat mask = cv::Mat::zeros(mosaic.size(), CV_8UC1);

    if (gridSize <= 0 || intensity < 0.01) return mask;

    //фиксированная ширина 1-3 пикселя
    int lineWidth = 1 + static_cast<int>(intensity * 2);

    //вертикальные швы
    for (int x = gridSize; x < mosaic.cols; x += gridSize) {
        int startX = std::max(0, x - lineWidth / 2);
        int endX = std::min(mosaic.cols, startX + lineWidth);

        cv::rectangle(mask,
            cv::Rect(startX, 0, endX - startX, mosaic.rows),
            cv::Scalar(255), -1);
    }

    //горизонтальные швы  
    for (int y = gridSize; y < mosaic.rows; y += gridSize) {
        int startY = std::max(0, y - lineWidth / 2);
        int endY = std::min(mosaic.rows, startY + lineWidth);

        cv::rectangle(mask,
            cv::Rect(0, startY, mosaic.cols, endY - startY),
            cv::Scalar(255), -1);
    }

    return mask;
}

//класс EffectFactory
//создает экземпляр эффекта постобработки по его имени
//возвращает unique_ptr на созданный эффект или nullptr для неизвестных эффектов
std::unique_ptr<PostProcessEffect> EffectFactory::createEffect(const std::string& effectName) {
    if (effectName == "color_correction") {
        return std::make_unique<ColorCorrectionEffect>();
    }
    else if (effectName == "alpha_blend") {
        return std::make_unique<AlphaBlendEffect>();
    }
    else if (effectName == "seam_smoothing") {
        return std::make_unique<SeamSmoothingEffect>();
    }
    else {
        return nullptr; //неизвестный эффект
    }
}

//класс PostProcessPipeline
//сборка постобработки
void PostProcessPipeline::setup(const PostProcessConfig& config) {
    effects.clear(); //очищаем предыдущие эффекты
    gridSize = config.gridSize; //сохраняем размер сетки

    //создаем и настраиваем каждый эффект из конфигурации
    for (const auto& [effectName, intensity] : config.effects) {
        auto effect = EffectFactory::createEffect(effectName);
        if (effect) {
            //для эффекта сглаживания швов дополнительно устанавливаем размер сетки
            if (auto seamEffect = dynamic_cast<SeamSmoothingEffect*>(effect.get())) {
                seamEffect->setGridSize(gridSize);
            }

            //добавляем эффект в цепочку
            effects.push_back(std::move(effect));
        }
    }
}

//применяет всю цепочку эффектов к изображению мозаики
//(эффекты применяются последовательно в порядке их добавления в конфигурации)
cv::Mat PostProcessPipeline::process(const cv::Mat& mosaic, const cv::Mat& original) {
    cv::Mat result = mosaic.clone();

    //последовательно применяем все эффекты из цепочки
    for (const auto& effect : effects) {
        result = effect->apply(result, original);
    }

    return result; //возвращаем финальный результат
}

File: post_processor.py
import cv2
import numpy as np
from typing import List, Tuple, Optional

class PostProcessConfig:
    def __init__(self):
        self.effects: List[Tuple[str, float]] = []
        self.grid_size: int = 30
    
    def add_effect(self, name: str, intensity: float = 0.5):
        self.effects.append((name, intensity))
    
    def clear_effects(self):
        self.effects.clear()


class PostProcessEffect:
    def apply(self, mosaic: np.ndarray, original: np.ndarray) -> np.ndarray:
        raise NotImplementedError
    
    def get_name(self) -> str:
        raise NotImplementedError
    
    def set_grid_size(self, grid_size: int):
        pass


class ColorCorrectionEffect(PostProcessEffect):
    def __init__(self, intensity: float = 0.5):
        self.intensity = intensity
    
    def apply(self, mosaic: np.ndarray, original: np.ndarray) -> np.ndarray:
        # Добавляем проверку на нулевую интенсивность
        if self.intensity <= 0.01:
            return mosaic.copy()
            
        corrected = mosaic.copy()
        original_resized = cv2.resize(original, (mosaic.shape[1], mosaic.shape[0]))
        
        mosaic_mean = cv2.mean(mosaic)
        original_mean = cv2.mean(original_resized)
        
        # Разделение на каналы
        mosaic_channels = cv2.split(mosaic)
        corrected_channels = []
        
        for i in range(len(mosaic_channels)):
            scale = original_mean[i] / (mosaic_mean[i] + 1e-5)
            scale = 1.0 + (scale - 1.0) * self.intensity
            
            corrected_channel = mosaic_channels[i].astype(np.float32) * scale
            corrected_channel = np.clip(corrected_channel, 0, 255).astype(np.uint8)
            corrected_channels.append(corrected_channel)
        
        corrected = cv2.merge(corrected_channels)
        return corrected
    
    def get_name(self) -> str:
        return "color_correction"


class AlphaBlendEffect(PostProcessEffect):
    def __init__(self, alpha: float = 0.5):
        self.alpha = alpha
    
    def apply(self, mosaic: np.ndarray, original: np.ndarray) -> np.ndarray:
        # Добавляем проверку на нулевую интенсивность
        if self.alpha <= 0.01:
            return mosaic.copy()
            
        original_resized = cv2.resize(original, (mosaic.shape[1], mosaic.shape[0]))
        blended = cv2.addWeighted(mosaic, 1.0 - self.alpha, original_resized, self.alpha, 0)
        return blended
    
    def get_name(self) -> str:
        return "alpha_blend"


class SeamSmoothingEffect(PostProcessEffect):
    def __init__(self, intensity: float = 0.5, grid_size: int = 30):
        self.intensity = intensity
        self.grid_size = grid_size
    
    def apply(self, mosaic: np.ndarray, original: np.ndarray) -> np.ndarray:
        # Добавляем проверку на нулевую интенсивность
        if self.intensity <= 0.01:
            return mosaic.copy()
            
        smoothed = mosaic.copy()
        seam_mask = self._create_seam_mask(mosaic)
        
        if cv2.countNonZero(seam_mask) == 0:
            return smoothed
        
        # Параметры размытия
        blur_size = 3 + int(self.intensity * 5)
        if blur_size % 2 == 0:
            blur_size += 1
        blur_size = max(3, min(blur_size, min(mosaic.shape[0], mosaic.shape[1])))
        
        if blur_size < 3:
            return mosaic.copy()
        
        sigma = self.intensity * 2.0
        
        # Размытие
        blurred = cv2.GaussianBlur(mosaic, (blur_size, blur_size), sigma)
        
        # Смешивание
        blend_factor = self.intensity * 0.7
        blended = cv2.addWeighted(mosaic, 1.0 - blend_factor, blurred, blend_factor, 0)
        
        # Копирование сглаженных областей - ИСПРАВЛЕННАЯ ЧАСТЬ
        # Используем numpy индексирование вместо copyTo
        mask_indices = seam_mask > 0
        smoothed[mask_indices] = blended[mask_indices]
        
        return smoothed
    
    def get_name(self) -> str:
        return "seam_smoothing"
    
    def set_grid_size(self, grid_size: int):
        self.grid_size = grid_size
    
    def _create_seam_mask(self, mosaic: np.ndarray) -> np.ndarray:
        mask = np.zeros(mosaic.shape[:2], dtype=np.uint8)
        
        if self.grid_size <= 0 or self.intensity < 0.01:
            return mask
        
        line_width = 1 + int(self.intensity * 2)
        
        # Вертикальные швы
        for x in range(self.grid_size, mosaic.shape[1], self.grid_size):
            start_x = max(0, x - line_width // 2)
            end_x = min(mosaic.shape[1], start_x + line_width)
            mask[:, start_x:end_x] = 255
        
        # Горизонтальные швы
        for y in range(self.grid_size, mosaic.shape[0], self.grid_size):
            start_y = max(0, y - line_width // 2)
            end_y = min(mosaic.shape[0], start_y + line_width)
            mask[start_y:end_y, :] = 255
        
        return mask


class EffectFactory:
    @staticmethod
    def create_effect(effect_name: str, intensity: float = 0.5) -> Optional[PostProcessEffect]:
        effects = {
            "color_correction": ColorCorrectionEffect,
            "alpha_blend": AlphaBlendEffect,
            "seam_smoothing": SeamSmoothingEffect
        }
        
        if effect_name in effects:
            effect_class = effects[effect_name]
            effect = effect_class(intensity)
            
            # Для эффекта сглаживания швов устанавливаем интенсивность
            if effect_name == "seam_smoothing":
                effect.intensity = intensity
                
            return effect
        return None


class PostProcessPipeline:
    def __init__(self):
        self.effects: List[PostProcessEffect] = []
        self.grid_size: int = 30
    
    def setup(self, config: PostProcessConfig):
        self.effects.clear()
        self.grid_size = config.grid_size
        
        for effect_name, intensity in config.effects:
            # Пропускаем эффекты с очень низкой интенсивностью
            if intensity <= 0.01:
                continue
                
            effect = EffectFactory.create_effect(effect_name, intensity)
            if effect:
                if isinstance(effect, SeamSmoothingEffect):
                    effect.set_grid_size(self.grid_size)
                self.effects.append(effect)
    
    def process(self, mosaic: np.ndarray, original: np.ndarray) -> np.ndarray:
        result = mosaic.copy()
        
        for effect in self.effects:
            try:
                result = effect.apply(result, original)
            except Exception as e:
                print(f"Error applying effect {effect.get_name()}: {e}")
                # В случае ошибки продолжаем с предыдущим результатом
                continue
        
        return result